File: leetcode/p00106/p00106.go
package p00106

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func buildTree(inorder []int, postorder []int) *TreeNode {
	memo := make(map[int]int, len(inorder))
	for idx, val := range inorder {
		memo[val] = idx
	}

	root, _ := helper(inorder, postorder, memo, len(postorder)-1, 0, len(postorder)-1)
	return root
}

func helper(inorder, postorder []int, memo map[int]int, rootIdx, startIdx, endIdx int) (*TreeNode, int) {
	if startIdx > endIdx {
		return nil, rootIdx
	}

	root := &TreeNode{Val: postorder[rootIdx]}
	rootIdx--
	root.Right, rootIdx = helper(inorder, postorder, memo, rootIdx, memo[root.Val]+1, endIdx)
	root.Left, rootIdx = helper(inorder, postorder, memo, rootIdx, startIdx, memo[root.Val]-1)

	return root, rootIdx
}

File: leetcode/p00106/p00106.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        self.rootIdx, memo = len(postorder)-1, {val: idx for idx, val in enumerate(inorder)}

        def helper(startIdx: int, endIdx: int) -> TreeNode:
            if startIdx > endIdx:
                return None
            
            root = TreeNode(postorder[self.rootIdx])
            self.rootIdx -= 1
            root.right = helper(memo[root.val]+1, endIdx)
            root.left = helper(startIdx, memo[root.val]-1)

            return root
        
        return helper(0, len(postorder)-1)
